ladybug_geometry/polyskel.py

```diff
@@ -0,0 +1,558 @@
+# coding=utf-8
+"""
+Implementation of the straight skeleton algorithm as described by Felkel and Obdržálek in their 1998 conference paper Straight skeleton implementation.
+"""
+
+from __future__ import division
+
+import logging
+import heapq
+from itertools import *
+from collections import namedtuple
+import operator
+
+# Geometry classes
+from ladybug_geometry.geometry2d.polygon import Polygon2D
+from ladybug_geometry.geometry2d.pointvector import Point2D, Vector2D
+from ladybug_geometry.geometry2d.line import LineSegment2D
+from ladybug_geometry.geometry2d.ray import Ray2D
+from ladybug_geometry import intersection2d
+
+log = logging.getLogger("__name__")
+
+
+# For Debugging
+
+class Debug:
+	def __init__(self, image):
+		if image is not None:
+			self.im = image[0]
+			self.draw = image[1]
+			self.do = True
+		else:
+			self.do = False
+
+	def line(self, *args, **kwargs):
+		if self.do:
+			self.draw.line(*args, **kwargs)
+
+	def rectangle(self, *args, **kwargs):
+		if self.do:
+			self.draw.rectangle(*args, **kwargs)
+
+	def show(self):
+		if self.do:
+			self.im.show()
+
+_debug = Debug(None)
+
+def set_debug(image):
+	global _debug
+	_debug = Debug(image)
+
+
+# Skeleton Code
+
+def _window(lst):
+	prevs, items, nexts = tee(lst, 3)
+	prevs = islice(cycle(prevs), len(lst)-1, None)
+	nexts = islice(cycle(nexts), 1, None)
+	return zip(prevs, items, nexts)
+
+def _determinant(a, b):
+	res = a.x*b.y - b.x*a.y
+	return res
+
+def _approximately_equals(a, b):
+	return a==b or (abs(a-b) <= max(abs(a), abs(b)) * 0.001)
+
+def _approximately_same(point_a, point_b):
+	return _approximately_equals(point_a.x, point_b.x) and _approximately_equals(point_a.y, point_b.y)
+
+def _normalize_contour(contour):
+	"""
+	Consumes list of x,y coordinate tuples and returns list of Point2Ds.
+
+	Args:
+		contour: list of x,y tuples from contour.
+	Return:
+		 list of Point2Ds of contour.
+	"""
+	contour = [Point2D(float(x), float(y)) for (x,y) in contour]
+	normed_contour = []
+	for prev, point, next in _window(contour):
+		normed_prev = Point2D.normalize(point-prev)
+		normed_next = Point2D.normalize(next-point)
+		if not (point == next or normed_prev == normed_next):
+			normed_contour.append(point)
+
+	return normed_contour
+
+
+class _SplitEvent(namedtuple("_SplitEvent", "distance, intersection_point, vertex, opposite_edge")):
+	__slots__ = ()
+	def __str__(self):
+		return "{} Split event @ {} from {} to {}".format(self.distance, self.intersection_point, self.vertex, self.opposite_edge)
+
+class _EdgeEvent(namedtuple("_EdgeEvent", "distance intersection_point vertex_a vertex_b")):
+	__slots__ = ()
+	def __str__(self):
+		return "{} Edge event @ {} between {} and {}".format(self.distance, self.intersection_point, self.vertex_a, self.vertex_b)
```

Comment:
First time I see subclassing namedtuple. Looks like a great idea actually.

For styling same comments here. Too long and use of ".